keymanager/src/crypto/kdf.rs

```diff
@@ -288,60 +331,195 @@ impl Kdf {
+                    None => break,
+                };
+
+                let prev_checksum = Self::load_checksum(storage, next_generation);
+                if prev_checksum != curr_checksum {
+                    let mut inner = self.inner.write().unwrap();
+                    inner.reset();
+                    return Err(KeyManagerError::StorageCorrupted.into());
+                }
+
+                curr_checksum = Self::checksum_master_secret(&secret, &curr_checksum);
+                next_generation += 1;
+            }
+        }
+
+        // If only one master secret is missing, try using stored proposal.
+        if next_generation == generation {
+            if let Some(secret) = Self::load_master_secret_proposal(storage) {
+                // Proposed secret is untrusted and needs to be verified.
+                let next_checksum = Self::checksum_master_secret(&secret, &curr_checksum);
+
+                if next_checksum == checksum {
+                    Self::store_master_secret(storage, &runtime_id, &secret, generation);
+                    Self::store_checksum(storage, curr_checksum, generation);
+
+                    curr_checksum = next_checksum;
+                    next_generation += 1;
+                }
+            }
+        }
+
+        // Load and replicate the missing master secrets in reverse order so that every secret
+        // is verified against the consensus checksum before being saved.
+        let mut last_checksum = checksum.clone();
+        for generation in (next_generation..=generation).rev() {
+            // Check the local storage first.
+            let secret = Self::load_master_secret(storage, &runtime_id, generation);
+            if let Some(secret) = secret {
+                // Previous checksum is untrusted and needs to be verified.
+                let prev_checksum = Self::load_checksum(storage, generation);
+                let next_checksum = Self::checksum_master_secret(&secret, &prev_checksum);
+
+                if next_checksum != last_checksum {
+                    let mut inner = self.inner.write().unwrap();
+                    inner.reset();
+                    return Err(KeyManagerError::StorageCorrupted.into());
+                }
+
+                last_checksum = prev_checksum;
+                continue;
+            }
+
+            // Master secret wasn't found and needs to be fetched from another enclave.
+            // Fetched values are untrusted and need to be verified.
+            let vs = master_secret_fetcher(generation)?;
+            let (secret, prev_checksum) = match vs.checksum.is_empty() {
+                true => (vs.secret, runtime_id.0.to_vec()),
+                false => (vs.secret, vs.checksum),
+            };
+            let next_checksum = Self::checksum_master_secret(&secret, &prev_checksum);
+
+            if next_checksum != last_checksum {
+                return Err(KeyManagerError::MasterSecretChecksumMismatch.into());
```

Comment:
Do I understand right that one bad node could result in the replication process aborting even if another node may have the correct secret? Maybe this validation should be done as part of the node selection logic in the master secret fetcher?

Reply:
A single bad secret can abort initialization. However, if the node is restarted or initialization is retried, replication will resume where it left off because all verified secrets had been stored locally.
Agree, that we could use another node to replicate the remaining secrets. But in that case a bad node would get away with it as these errors would silently be ignored.
Took the comment into account and changed the fetcher. Now we silently ignore bad secrets.